internal/transports/p2p/peer/peer.go

```diff
@@ -0,0 +1,546 @@
+		if err != nil {
+			return err
+		}
+	}
+
+	p.queueMessage(msg)
+	return nil
+}
+
+func (p *Peer) readVerOrVerAck() error {
+	remoteMsg, _, err := wire.ReadMessage(p.conn, p.protocolVersion, p.chainParams.Net)
+	if err != nil {
+		p.log.Error().Msgf("could not read message from peer, reason: %v", err)
+		return err
+	}
+
+	switch msg := remoteMsg.(type) {
+	case *wire.MsgVersion:
+		err = p.handleVersionMessage(msg)
+		if err != nil {
+			return err
+		}
+	case *wire.MsgVerAck:
+		p.verAckReceived = true
+	default:
+		_, err = p.requireVerAckReceived(remoteMsg)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (p *Peer) handleVersionMessage(msg *wire.MsgVersion) error {
+	// Detect self connections.
+	if msg.Nonce == p.nonce {
+		err := errors.New("disconnecting peer connected to self")
+		p.log.Error().Msg(err.Error())
+		return err
+	}
+
+	p.protocolVersion = min(p.protocolVersion, uint32(msg.ProtocolVersion))
+	p.services = msg.Services
+	p.lastBlock = msg.LastBlock
+	p.startingHeight = msg.LastBlock
+	p.timeOffset = msg.Timestamp.Unix() - time.Now().Unix()
+	p.userAgent = msg.UserAgent
+
+	if uint32(msg.ProtocolVersion) < minAcceptableProtocolVersion {
+		reason := fmt.Sprintf("protocol version must be %d or greater", minAcceptableProtocolVersion)
+		p.writeRejectMessage(msg, reason)
+		p.log.Error().Msgf("%s, rejecting connection", reason)
+		return errors.New(reason)
+	}
+
+	return nil
+}
+
+func (p *Peer) requireVerAckReceived(remoteMsg wire.Message) (*wire.MsgVerAck, error) {
+	msg, ok := remoteMsg.(*wire.MsgVerAck)
+	if !ok {
+		reason := "missing-verack message, misbehaving peer"
+		p.writeRejectMessage(msg, reason)
+		p.log.Error().Msgf("%s, rejecting connection", reason)
+		return nil, errors.New(reason)
+	}
+	return msg, nil
+}
+
+func (p *Peer) handlePingMsg(msg *wire.MsgPing) {
+	p.log.Info().Msgf("received ping from peer %s with nonce: %d", p, msg.Nonce)
+	if p.protocolVersion > wire.BIP0031Version {
+		p.queueMessage(wire.NewMsgPong(msg.Nonce))
+	}
+}
+
+func (p *Peer) handleInvMsg(msg *wire.MsgInv) {
+	p.log.Info().Msgf("received inv msg from peer %s", p)
+	if !p.synced {
+		p.log.Info().Msgf("we are still syncing, ignoring inv msg from peer %s", p)
+		return
+	}
+
+	lastBlock := searchForFinalBlockIndex(msg.InvList)
+	if lastBlock == -1 {
+		p.log.Info().Msgf("no blocks in inv msg from peer %s", p)
+		return
+	}
+
+	// if last header from inv msg is already in database, ignore
+	lastBlockHash := &msg.InvList[lastBlock].Hash
+	_, err := p.headersService.GetHeightByHash(lastBlockHash)
+	if err == nil {
+		p.log.Info().Msgf("blocks from inv msg from peer %s already existsing in db", p)
+		return
+	}
+
+	p.log.Info().Msgf("requesting new headers from peer %s", p)
+	err = p.writeGetHeadersMsg(lastBlockHash)
+	if err != nil {
+		p.log.Error().Msgf("error while requesting new headers from peer %s, reason: %v", p, err)
+	}
+}
+
+func (p *Peer) handleHeadersMsg(msg *wire.MsgHeaders) {
```

Comment:
I moved the last part to a separate function, other than that, I don't really see a good way of refactoring it, other than throwing some comments, which are unncessary and we agreed to not use them. 

For me this code is pretty self-explanatory, but maybe it's because I've been sitting with it (and with the diagrams) for too long 😄

Reply:
Now it looks better, but I think that I still have some idea how to make it a little bit more clear.
If you don't mind, I could do a separate PR after merging this one.